cli/azd/pkg/tools/kubectl/kubectl.go

```diff
@@ -0,0 +1,265 @@
+func (cli *kubectlCli) CheckInstalled(ctx context.Context) (bool, error) {
+	return tools.ToolInPath("kubectl")
+}
+
+// Returns the installation URL to install the K8s CLI
+func (cli *kubectlCli) InstallUrl() string {
+	return "https://aka.ms/azure-dev/kubectl-install"
+}
+
+// Gets the name of the Tool
+func (cli *kubectlCli) Name() string {
+	return "kubectl"
+}
+
+// Sets the env vars available to the CLI
+func (cli *kubectlCli) SetEnv(envValues map[string]string) {
+	cli.env = envValues
+}
+
+// Sets the current working directory
+func (cli *kubectlCli) Cwd(cwd string) {
+	cli.cwd = cwd
+}
+
+// Sets the k8s context to use for future CLI commands
+func (cli *kubectlCli) ConfigUseContext(ctx context.Context, name string, flags *KubeCliFlags) (*exec.RunResult, error) {
+	res, err := cli.Exec(ctx, flags, "config", "use-context", name)
+	if err != nil {
+		return nil, fmt.Errorf("failed setting kubectl context: %w", err)
+	}
+
+	return &res, nil
+}
+
+// Views the current k8s configuration including available clusters, contexts & users
+func (cli *kubectlCli) ConfigView(
+	ctx context.Context,
+	merge bool,
+	flatten bool,
+	flags *KubeCliFlags,
+) (*exec.RunResult, error) {
+	kubeConfigDir, err := getKubeConfigDir()
+	if err != nil {
+		return nil, err
+	}
+
+	args := []string{"config", "view"}
+	if merge {
+		args = append(args, "--merge")
+	}
+	if flatten {
+		args = append(args, "--flatten")
+	}
+
+	runArgs := exec.NewRunArgs("kubectl", args...).
+		WithCwd(kubeConfigDir).
+		WithEnv(environ(cli.env))
+
+	res, err := cli.executeCommandWithArgs(ctx, runArgs, flags)
+	if err != nil {
+		return nil, fmt.Errorf("kubectl config view: %w", err)
+	}
+
+	return &res, nil
+}
+
+func (cli *kubectlCli) ApplyPipe(ctx context.Context, input string, flags *KubeCliFlags) (*exec.RunResult, error) {
+	runArgs := exec.
+		NewRunArgs("kubectl", "apply", "-f", "-").
+		WithEnv(environ(cli.env)).
+		WithStdIn(strings.NewReader(input))
+
+	res, err := cli.executeCommandWithArgs(ctx, runArgs, flags)
+	if err != nil {
+		return nil, fmt.Errorf("kubectl apply -f: %w", err)
+	}
+
+	return &res, nil
+}
+
+// Applies manifests from the specified input
+func (cli *kubectlCli) ApplyFiles(ctx context.Context, path string, flags *KubeCliFlags) error {
+	entries, err := os.ReadDir(path)
+	if err != nil {
+		return fmt.Errorf("failed reading files in path, '%s', %w", path, err)
+	}
+
+	for _, entry := range entries {
+		if entry.IsDir() {
+			continue
+		}
+
+		ext := filepath.Ext(entry.Name())
+		if !(ext == ".yaml" || ext == ".yml") {
+			continue
+		}
+
+		filePath := filepath.Join(path, entry.Name())
+		fileBytes, err := os.ReadFile(filePath)
+		if err != nil {
+			return fmt.Errorf("failed reading manifest file '%s', %w", filePath, err)
+		}
+
+		yaml := string(fileBytes)
+		replaced, err := envsubst.Eval(yaml, func(name string) string {
+			if val, has := cli.env[name]; has {
+				return val
+			}
+			return os.Getenv(name)
+		})
```

Comment:
We may need to ship with this for now and then we can roll this back after we add support for Kustomize and/or helm since they make it easier to inject configuration along with manifests.

Reply:
This also makes me nervous.  I was relieved when we moved towards doing selective substitutions in `azure.yaml` because I was always scared about cases like this.

The problem doing that here is that these files sort of have arbitrary schemas, so it's hard for us to understand which nodes should be "replacement strings" and which should be "normal strings".

Maybe we should think about something built around text/template?  Like as a strawman, if you have `.ymal.template` (or `.yml.template)` we'll do a text/template replacement to generate the file we'd apply?  That might feel more natural to k8s folks and might be a decent model and less likely to conflict with existing embeded stuff?

Tactically, I think I'm /okay/ if we land with this as-is, but we should file a follow up issue to track doing the right thing here and then use that as a gate for us feeling at "GA" quality about the k8s deployment target.